lib/rucio/rse/protocols/gfal.py

```diff
@@ -436,14 +436,19 @@ def __gfal2_copy(self, src, dest, src_spacetoken=None, dest_spacetoken=None, tra
         :raises RucioException: if it failed to copy the file.
         """
         ctx = self.__ctx
-        params = ctx.transfer_parameters()
```

Comment:
What is your motivation for moving this line?

Reply:
Just to set the context parameters before

Reply:
Forgive me, I still do not understand.

It seems to me that there might be some unexpected quirks or hidden side-effects in the GFAL interface that one cannot anticipate just by reading the source code alone. I think we must document them *somewhere* (e.g. in the source code, the commit message, or the issue), otherwise future developments may silently introduce a regression.

Essentially, I have two questions:

```python
            params.timeout = int(transfer_timeout)
```

Why is this not sufficient? Why is it necessary to set GFAL plug-in timeouts?

```python
            ctx.set_opt_integer("HTTP PLUGIN", "OPERATION_TIMEOUT", transfer_timeout)
            ctx.set_opt_integer("SRM PLUGIN", "OPERATION_TIMEOUT", transfer_timeout)
            ctx.set_opt_integer("GRIDFTP PLUGIN", "OPERATION_TIMEOUT", transfer_timeout)
            params = ctx.transfer_parameters()
            params.timeout = int(transfer_timeout)
```

Why is the order important? Why does one *have* to first set the plug-in timeouts, then the transfer timeout? If it is not, then I would have found it preferable to merely introduce the context configuration after line 444.

Reply:
The first one I wondered about too when I first looked at this issue. My (pure speculation based) interpretation was that there is a difference in gfal between a timeout and an OPERATION_TIMEOUT, maybe the timeout gets only triggered when data is already flowing, and the OPERATION_TIMEOUT happens already during the connection-phase.

The second one, I agree, I would move this back up, simply to also get rid of this rather ugly else block :-)

If you can do it before the afternoon, that would be great for the release.